src/SAWScript/Proof.hs

```diff
@@ -1141,56 +1825,109 @@ tacticIntro :: (F.MonadFail m, MonadIO m) =>
+             body' <- liftIO $ instantiateVar sc 0 x body
+             let goal' = goal { goalSequent = mkSqt (Prop body') }
+             return (tt, mempty, [goal'], introEvidence xv)
+
+        _ -> fail "intro tactic failed: not a function"
+
+    _ -> fail "intro tactic: conclusion focus required"
+
+
+tacticIntroHyps :: (F.MonadFail m, MonadIO m) => SharedContext -> Integer -> Tactic m ()
+tacticIntroHyps sc n = Tactic \goal ->
+  case goalSequent goal of
+    GoalFocusedSequent hs (FB gs1 g gs2) ->
+      do (newhs, g') <- liftIO (loop n g)
+         let sqt' = GoalFocusedSequent (hs ++ newhs) (FB gs1 g' gs2)
+         let goal' = goal{ goalSequent = sqt' }
+         return ((), mempty, [goal'], updateEvidence (NormalizeSequentEvidence sqt'))
+    _ -> fail "goal_intro_hyps: conclusion focus required"
+
+ where
+   loop i g
+     | i <= 0 = return ([],g)
+     | otherwise =
+         splitImpl sc g >>= \case
+           Nothing -> fail "intro_hyps: could not find enough hypotheses to introduce"
+           Just (h,g') ->
+             do (hs,g'') <- loop (i-1) g'
+                return (h:hs, g'')
+
+tacticRevertHyp :: (F.MonadFail m, MonadIO m) => SharedContext -> Integer -> Tactic m ()
+tacticRevertHyp sc i = Tactic \goal ->
+  case goalSequent goal of
+    GoalFocusedSequent hs (FB gs1 g gs2) ->
+      case genericDrop i hs of
+        (h:_) ->
+          case (asEqTrue (unProp h), asEqTrue (unProp g)) of
+            (Just h', Just g') ->
+              do g'' <- liftIO (Prop <$> (scEqTrue sc =<< scImplies sc h' g'))
+                 let sqt' = GoalFocusedSequent hs (FB gs1 g'' gs2)
+                 let goal' = goal{ goalSequent = sqt' }
+                 return ((), mempty, [goal'], updateEvidence (NormalizeSequentEvidence sqt'))
+
+            _ -> fail "goal_revert_hyp: expected EqTrue props"
+        _ -> fail "goal_revert_hyp: not enough hypotheses"
+    _ -> fail "goal_revert_hyp: conclusion focus required"
+
+
+-- | Attempt to prove a goal by applying a local hypothesis.  Any hypotheses of
+--   the applied proposition will generate additional subgoals.
+tacticApplyHyp :: (F.MonadFail m, MonadIO m) => SharedContext -> Integer -> Tactic m ()
+tacticApplyHyp sc n = Tactic \goal ->
+  case goalSequent goal of
+    UnfocusedSequent{} -> fail "apply hyp tactic: focus required"
+    HypFocusedSequent{} -> fail "apply hyp tactic: cannot apply in a hypothesis"
+    GoalFocusedSequent hs (FB gs1 g gs2) ->
+      case genericDrop n hs of
+        (h:_) ->
+          liftIO (goalApply sc h g) >>= \case
+            Nothing -> fail "apply hyp tactic: no match"
+            Just newterms ->
+              let newgoals =
+                    [ goal{ goalSequent = GoalFocusedSequent hs (FB gs1 p gs2)
+                          , goalType = goalType goal ++ ".subgoal" ++ show i
+                          }
+                    | Right p <- newterms
+                    | i <- [0::Integer ]
+                    ] in
```

Comment:
The use of `| i <- [0::Integer ]` here suggests that `newgoals` will only ever be of length zero or one. Is that right?

Reply:
ohhh... that doesn't look right, it should be [ 0 .. ], like in tacticApply